fuzzers/main.py

```diff
@@ -0,0 +1,949 @@
+            log_lines = list(log_file_r)
+        branch = self.branch['name']
+        logs_url = f'https://storage.cloud.google.com/fuzzer/logs/{self.log_relpath}'
+
+        # Identify the artifact path
+        artifact_pattern = f'Test unit written to {corpus.artifacts_for(self.target)}/'
+        artifact = '<failed detecting relevant artifact in log file>'
+        for line in log_lines[::-1]:
+            if artifact_pattern in line:
+                artifact = line.split(artifact_pattern)[1].strip()
+                break
+
+        # Check the artifact was not reported yet for this branch
+        if artifact in REPORTED_ARTIFACTS[branch]:
+            return
+        already_reported_for = [
+            other_branch for (other_branch,
+                              reported_artifacts) in REPORTED_ARTIFACTS.items()
+            if artifact in reported_artifacts
+        ]
+        REPORTED_ARTIFACTS[branch].append(artifact)
+
+        if already_reported_for:
+            already_reported_msg = f'(already reported for branches {already_reported_for})'
+        else:
+            already_reported_msg = ''
+
+        # Send the information in two mesages, a short one guaranteed to succeed, then a long one
+        # with the log lines that could go over the message size limit
+        client = zulip.Client(config_file=ZULIPRC)
+        gcs_artifact = shlex.quote(
+            f'gs://{bucket.name}/{self.corpus_vers}/{self.target["crate"]}/'
+            f'{self.target["runner"]}/artifacts/{artifact}')
+        quoted_crate = shlex.quote(self.target['crate'])
+        quoted_runner = shlex.quote(self.target['runner'])
+        quoted_artifact = shlex.quote(
+            f'artifacts/{self.target["runner"]}/{artifact}')
+        client.send_message({
+            'type':
+                'stream',
+            'to':
+                'pagoda/fuzzer/private',
+            'topic':
+                f'{branch}: artifact {artifact}',
+            'content':
+                f'''\
+# Fuzzer found new crash for branch *{branch}* {already_reported_msg}
+
+Full logs are available at {logs_url}.
+
+To reproduce, first go to the fuzzer directory:
+```
+cd {quoted_crate}
+```
+
+You can download the artifact by using the following command:
+```
+gsutil cp {gcs_artifact} {quoted_artifact}
+```
+
+Then, you can reproduce by running the following command:
+```
+RUSTC_BOOTSTRAP=1 cargo fuzz run {quoted_runner} {quoted_artifact}
+```
+
+Or minimize by running the following command:
+```
+RUSTC_BOOTSTRAP=1 cargo fuzz tmin {quoted_runner} {quoted_artifact}
+```
+
+Please edit the topic name to add more meaningful information once investigated. \
+Keeping the artifact hash in it can help if the same artifact gets detected as \
+crashing another branch.
+'''
+        })
+        prepend_log_lines = '\n'.join(log_lines[:5]) + '\n[...]\n'
+        focus_log_lines = ''
+        for line in log_lines[::-1]:
```

Comment:
You're right that I missed the fact that the log lines would be duplicated, this is now fixed :)

About unifying with read_short_log, I think the use cases are different enough to warrant two different functions: first, because this code also needs to censor the triple-` that marks the end of spoiler elements, but also because read_short_log takes half-and-half whereas for fuzzing only the end is important, as all the beginning is nothing but repeated messages… except for the first 5 lines that we output ourselves that have the environmental metadata about the fuzzer (the part for which you submitted a code suggestion above)